### PR TITLE
[PRE-DRAFT] Hypermedia Link PSR

### PR DESCRIPTION
**Pull request by Crell**
Proposed PSR for Hypermedia Link handling.  Currently in pre-draft state.  Merge only when it passes an entrance vote.

### CODE REVIEW

File: proposed/links-meta.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+This specification aims to provide PHP developers with a simple, common way of representing a
+hypermedia link independently of the serialization format that is used. That in turn allows
+a system to serialize a response with hypermedia links into one or more wire formats independently
+of the process of deciding what those links should be.
+
+### Open questions
+
+The following questions are still outstanding, in the opinion of the Editor, and should be resolved.
+
+* LinkableInterface is a terrible name. Please suggest another one.
```
Comment (samdark) [inline proposed/links-meta.md]: Isn't it LinkInterface now?

File: proposed/links.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+ * 
+ */
+interface LinkableInterface
+{
+    /**
+     * Returns a collection of LinkInterface objects.
+     *
+     * The collection may be an array or any PHP \Traversable object.
+     *
+     * @return array|\Traversable
```
Comment (localheinz) [inline proposed/links.md]: @Crell 

How about using

```php
/**
 * @return LinkInterface[]|\Traversable
*/
```

here?

File: proposed/links.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+ * 
+ */
+interface LinkInterface
+{
+    /**
+     * Returns the target of the link.
+     *
+     * The target must be a URI or a Relative URI reference.
+     *
+     * @return string
```
Comment (localheinz) [inline proposed/links.md]: @Crell 

I'm also :+1: for returning an instance of [`Psr\Http\Message\UriInterface`](https://github.com/php-fig/http-message/blob/master/src/UriInterface.php) here!

File: proposed/links.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+     * Optionally the microformats.org registry may be used, but this may not
+     * be valid in every context:
+     *
+     * http://microformats.org/wiki/existing-rel-values
+     *
+     * Private relationship types should always be an absolute URI.
+     *
+     * @return string[]
+     */
+    public function getRel();
```
Comment (localheinz) [inline proposed/links.md]: @Crell 

Can we afford a little more expressiveness, or does it really make more sense to name the accessor after the `<link>` attribute?

File: proposed/links.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+     * encoding the values they support.
+     *
+     * Any value that appears that is not valid in the context in which it is
+     * used should be ignored.
+     *
+     * Some attributes, (commonly hreflang) may appear more than once in their
+     * context. Attributes such as those may be specified as an array of
+     * strings.
+     *
+     * @return array
```
Comment (localheinz) [inline proposed/links.md]: @Crell 

How about using

```php
/**
 * @return string[]
*/
```

here?

File: proposed/links.md [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+#### 3.2.1 `Psr\Http\Message\LinkableInterface`
+
+```php
+<?php
+namespace Psr\Http\Link;
+
+/**
+ * 
+ */
+interface LinkableInterface
```
Comment (localheinz) [inline proposed/links.md]: @Crell 

Would `Psr\Http\Link\ResourceInterface` make more sense?

File: proposed/links.md [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+     * context. Attributes such as those may be specified as an array of
+     * strings.
+     *
+     * @return array
+     */
+    public function getAttributes();
+}
+```
+
+#### 3.2.1 `Psr\Http\Message\LinkableInterface`
```
Comment (renan) [inline proposed/links.md]: Should be `Psr\Http\Link\LinkableInterface` according to namespace

**Comment by Crell** [comment]
This passed its entrance vote, so merging.  I'll address the comments above later, once this is merged.

File: proposed/links.md (hunk 6 above)
Comment (quimcalpe) [inline proposed/links.md]: @Crell `LinkCollectionInterface` would be the more obvious to me in this context, but URIs, URLs and URNs are references at the end of the day, so `ReferenceListInterface` or `ReferenceCollectionInterface` could also make sense.

Comment (michaelcullum) [inline proposed/links.md]: @quimcalpe Best place to discuss this is the mailing list: https://groups.google.com/forum/#!topic/php-fig/SFJK_Ynivu4

**Pull request merged**